InteractionBase/src/vis/ViewSwitcherMenu.cpp

```diff
@@ -0,0 +1,141 @@
+ **    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+#include "ViewSwitcherMenu.h"
+
+#include "nodes/ViewSwitcherEntry.h"
+#include "vis/VViewSwitcherEntry.h"
+#include "VisualizationBase/src/items/ViewItem.h"
+#include "VisualizationBase/src/cursor/TextCursor.h"
+
+namespace Interaction {
+
+ITEM_COMMON_DEFINITIONS(ViewSwitcherMenu, "item")
+
+ViewSwitcherMenu* ViewSwitcherMenu::instance{};
+
+void ViewSwitcherMenu::show(QVector<Model::Node*> selectableNodes, Visualization::Item* target)
+{
+	QApplication::postEvent(target->scene(),
+							new Visualization::CustomSceneEvent( [=]() { showNow(selectableNodes, target); }));
+}
+
+void ViewSwitcherMenu::hide()
+{
+	if (instance)
+		QApplication::postEvent(instance->scene(),
+								new Visualization::CustomSceneEvent( [&]() { hideNow(); }));
+}
+
+void ViewSwitcherMenu::showNow(QVector<Model::Node*> selectableNodes, Visualization::Item* target)
+{
+	ViewSwitcherMenu::hideNow();
+	instance = new ViewSwitcherMenu(selectableNodes, target);
+	target->scene()->addTopLevelItem(instance);
+	instance->installSceneEventFilter(instance);
+	instance->setFiltersChildEvents(true);
+}
+
+void ViewSwitcherMenu::hideNow()
+{
+	if (instance)
+	{
+		instance->removeSceneEventFilter(instance);
+		instance->scene()->removeTopLevelItem(instance);
+	}
+	SAFE_DELETE_ITEM(instance);
+	instance = nullptr;
+}
+
+ViewSwitcherMenu::ViewSwitcherMenu(QVector<Model::Node*> selectableNodes,
+											 Visualization::Item* target, StyleType* style)
+	: Super(selectableNodes, target, style)
+{
+}
+
+bool ViewSwitcherMenu::sceneEventFilter(QGraphicsItem *watched, QEvent *event)
+{
+	if (inEditMode_ && event->type() == QEvent::KeyPress)
+	{
+		//If editing, let the mover use the cursor keys to move within the item itself,
+		//rather than between items
+		auto keyEvent = static_cast<QKeyEvent*>(event);
```

Comment:
I feel like this should somehow be part of the base functionality. For example, if we just use the arrow keys we select different menu items. But if we press some special key, we enter "menu entry focus mode" which forwards all subsequent events to that menu entry directly.